pkg/ocibundle/tools/overlay.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package tools
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+func deleteDir(dir string, err error) {
+	if err != nil {
+		os.RemoveAll(dir)
+	}
+}
+
+// CreateOverlay creates a writable overlay
+func CreateOverlay(bundlePath string) (err error) {
+	oldumask := syscall.Umask(0)
+	defer syscall.Umask(oldumask)
+
+	overlayDir := filepath.Join(bundlePath, "overlay")
+	if err = os.Mkdir(overlayDir, 0700); err != nil {
+		return
+	}
+	// delete overlay directory in case of error
+	defer deleteDir(overlayDir, err)
+
+	if syscall.Mount(overlayDir, overlayDir, "", syscall.MS_BIND, ""); err != nil {
+		return err
+	}
+	// best effort to cleanup mount
+	defer func() {
+		if err != nil {
+			syscall.Unmount(overlayDir, syscall.MNT_DETACH)
+		}
+	}()
+
+	if syscall.Mount("", overlayDir, "", syscall.MS_REMOUNT|syscall.MS_BIND, ""); err != nil {
+		return err
+	}
+
+	upperDir := filepath.Join(overlayDir, "upper")
+	if err = os.Mkdir(upperDir, 0700); err != nil {
+		return
+	}
+	workDir := filepath.Join(overlayDir, "work")
+	if err = os.Mkdir(workDir, 0700); err != nil {
+		return
```

Comment:
I am not a fan of these empty `if err ! = nil { return err }` type error passthroughs. If these ever error and we need to debug I have a feeling it will be incredibly hard to debug this because we don't know where the error is definitely coming from. I would much prefer: `return fmt.Errorf("while making overlay workDir (%s): %s", workDir, err)`

returning errors in this way should be much easier too debug should that need arise